cas-mfa-java/src/main/java/net/unicon/cas/mfa/authentication/RegisteredServiceMfaRoleProcessor.java

```diff
@@ -0,0 +1,254 @@
+package net.unicon.cas.mfa.authentication;
+
+import net.unicon.cas.addons.serviceregistry.RegisteredServiceWithAttributes;
+import net.unicon.cas.mfa.web.support.MultiFactorAuthenticationSupportingWebApplicationService;
+import net.unicon.cas.mfa.web.support.MultiFactorWebApplicationServiceFactory;
+import org.jasig.cas.authentication.Authentication;
+import org.jasig.cas.authentication.principal.WebApplicationService;
+import org.jasig.cas.services.RegisteredService;
+import org.jasig.cas.services.ServicesManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.validation.constraints.NotNull;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * Resolves potential mfa request based on the configured principal attribute and the service attribute.
+ * If a service has an mfa-role attribute, then its authn_method is only enforce if the principal matches the requested attribute.
+ *
+ * @author John Gasper
+ * @author Unicon, inc.
+ */
+
+public class RegisteredServiceMfaRoleProcessor {
```

Comment:
This is fine, but you should likely implement an interface to allow for pluggable impls.